package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,331 @@
+
+    """
+
+    def __init__(self, u, select=None, msd_type='xyz', fft=True, **kwargs):
+        r"""
+        Parameters
+        ----------
+        u : Universe
+            An MDAnalysis :class:`Universe`.
+        selection : str
+            An MDAnalysis selection string. Defaults to `None`.
+        msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Desired dimensions to be included in the MSD. Defaults to 'xyz'.
+        fft : bool
+            Use a fast FFT based algorithm for computation of the MSD.
+            Otherwise, use the simple "windowed" algorithm. Defaults to ``True``.
+
+        """
+        self.u = u
+        
+        super(EinsteinMSD, self).__init__(self.u.trajectory, **kwargs)
+
+        #args
+        self.select = select
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self._dim = None
+        self._position_array = None
+        self._atoms = None
+        
+        #indexing
+        self._n_frames = 0
+        self._n_particles = 0
+
+        #result
+        self.dim_fac = 0
+        self.timeseries = None
+        self.msds_by_particle = None
+
+    def _prepare(self):
+        self._parse_msd_type()
+        self._atoms = self.u.select_atoms(self.select)
+        self._n_frames = len(self.u.trajectory)
+        self._n_particles = len(self._atoms)
+        self._position_array = np.zeros((self._n_frames, self._n_particles, 3))
+        self.msds_by_particle = np.zeros((self._n_frames, self._n_particles))
+        # self.timeseries not set here
+        
+    def _parse_msd_type(self):
+        r""" Sets up the desired dimensionality of the MSD.
+        
+        Parameters
+        ----------
+        self.msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Dimensions to be included in the MSD.
+
+        Returns
+        -------
+        self._dim : list
+            Array-like used to slice the positions to obtain desired dimensionality
+        self.dim_fac : int
+            Dimension factor :math:`d` of the MSD.
+
+        """
+        keys = {'x':[0], 'y':[1], 'z':[2], 'xy':[0,1], 'xz':[0,2], 'yz':[1,2], 'xyz':[0,1,2]}
+
+        self.msd_type = self.msd_type.lower()
+
+        try:
+            self._dim = keys[self.msd_type]
+        except KeyError:
+            raise ValueError('invalid msd_type: {} specified, please specify one of xyz, xy, xz, yz, x, y, z'.format(self.msd_type))
+
+        self.dim_fac = len(self._dim)
+
+    def _single_frame(self):
+        r""" Constructs array of positions for MSD calculation.
+        
+        Parameters
+        ----------
+        self.u : :class:`Universe`
+            MDAnalysis Universe
+        
+        Returns
+        -------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape = (n_frames, n_particles, 3)
+        """
+
+        self._position_array[self._frame_index,:,:] = self._atoms.positions
```

Comment:
Relates to keeping `_position_array` as small as possible above.

```suggestion
        self._position_array[self._frame_index] = self._atoms.positions[:, self._dim]
```